Refactor demo for py-allspice

danielallspice opened the pull request:
Refactored helloworld to work with py-allspice
Added new example for allspice_generated/json
Removed most of the other examples to rewrite them with better patterns in future branches. I wanted a clean example for main.

Fixes #5

danielallspice committed:
Copy example project to start

danielallspice committed:
Remove files

Bad project start

danielallspice committed:
Add basic repo json example

danielallspice committed:
Add getrepoJSON

danielallspice committed:
Update to show usage

danielallspice committed:
Add example output file

danielallspice committed:
Remove old comments

jtran commented on GetRepoJSON/getRepoJSON.py:
```diff
@@ -0,0 +1,52 @@
+
+import sys
+sys.path.append('..')
+from allspice_proto.allspice_proto import AllSpice_Proto
```
```suggestion
from allspice_proto.allspice_proto import AllspiceProto
```

Convention in Python is that class names are camel case.

If I'm being pedantic, "allspice" is one word, which is why the S should be lowercase.  Everyone agrees that in snake case, it's `allspice_proto`, but I haven't been able to get people to agree on what the same thing is in camel case.

danielallspice replied:
I changed all code references to "allspice". I left non-code branded references to "AllSpice".

jtran commented on GetRepoJSON/getRepoJSON.py:
```diff
@@ -0,0 +1,52 @@
+# -----------------------------------------------------------------------------
+allspice.infoheader("Test GET /version")
+# test GET server version
+versionResponse = ""
```
Variables should be snake case.

```suggestion
version_response = ""
```

danielallspice replied:
I converted all the local variables to snake case.

jtran commented on pygitea_mod/gitea/gitea.py:
```diff
@@ -416,4 +426,44 @@ def modify_team(
+        )
+
+        
+        pass
```
```suggestion
```

jtran commented on pygitea_mod/gitea/gitea.py:
```diff
@@ -416,4 +426,44 @@ def modify_team(
         return api_object
+    
+
+    def getRepoJSON(
```
```suggestion
    def get_repo_json(
```

danielallspice replied:
Removed all camel case function names.

jtran commented on pygitea_mod/gitea/gitea.py:
```diff
@@ -157,6 +159,14 @@ def requests_patch(self, endpoint: str, data: dict):
+        request = self.requests.patch(self.__get_url(endpoint), headers=self.headers, data=json.dumps(data))
+        if request.status_code not in [200, 201]:
+            error_message = f"Received status code: {request.status_code} ({request.url}) {data}"
+            self.logger.error(error_message)
+            raise Exception(error_message)
```
Log-and-raise is an anti-pattern.  Do one or the other.  (I would probably raise.)  If you do both, presumably something else at a higher level is going to catch the exception and do something with it, probably print or log it.  So you end up with duplicates.

danielallspice replied:
Thank you. I have all sorts of questions on error checking, try/except blocks.

jtran commented on GetRepoJSON/getRepoJSON.py:
```diff
@@ -0,0 +1,52 @@
+
+from time import sleep
+def delayserver():
+    sleep(0.01)
```
😢 I assume this is here to deal with rate limiting?

danielallspice replied:
This is rate limiting. I could use suggestions. Whatever I put here, the customer is going to do too.

jtran replied:
Basically, you have one request executor that all requests go through.  It keeps track of how long it's been since the last request (per host -- in case you're hitting multiple APIs, each with its own limit), how many requests are going simultaneously, etc., and it can do its own client-side rate limiting.  If you're under the limit, there's no need to wait, and everything can be faster.  I would have to think about it more since I don't know of an off-the-shelf library that does this.  Maybe it exists though.  I haven't needed it before.  I think we could write our own too.

danielallspice replied:
I would like to move the rate limiting to our python module to limit user error. https://github.com/AllSpiceIO/py-allspice/issues/6

jtran commented on GetRepoJSON/getRepoJSON.py:
```diff
@@ -0,0 +1,52 @@
+
+
+import sys
+sys.path.append('..')
```
😧 I understand why this is here, but this causes me anguish.  Probably not worth dealing with it for these scripts.  Python modules are stupidly complicated.

danielallspice replied:
I actually had a question about the best way to do this. Some of the functionality I can move to py-allspice, but I wasn't sure the best way to share common functions in a python project. I'm happy to rewrite if you know of a better way.

jtran replied:
When you run a script with `python foo/bar.py`, the directory `foo` gets added to the path.  On the other hand, if you run the script with `python -m foo.bar` from the repo root, the current working directory, the repo root in this case, gets added to the path.  It is also aware of its package, so relative imports can work.  Even though both ways run the same Python script, only the latter leads to intuitive imports (both absolute and relative) that work consistently, whether installed with pip or not, without mucking with `sys.path`.

A way to work around this is to only ever run scripts at the top level.  The top-level scripts can import and call whatever nested module they want.  Then the two are equivalent.  This is what we do in AllSpice-Tools.  But I understand that this can be unsatisfactory.  One could argue that the habit of running things with `python foo/bar.py` should be stopped in favor of `python -m foo.bar`.  However, that would also mean that you can't run the file directly as in `./foo/bar.py` with a shebang, and shell completion no longer works.  That's a bummer.

danielallspice replied:
I removed this reference. The new py-allspice module removes the reason to use this.

kdumontnu commented:
I would really like to have your changes to pygitea ported here: https://github.com/AllSpiceIO/py-allspice

So that we don't have to do double-work and ask customers to download the custom source here.

danielallspice commented:
> I would really like to have your changes to pygitea ported here: https://github.com/AllSpiceIO/py-allspice
> 
> So that we don't have to do double-work and ask customers to download the custom source here.

Sounds great. It makes sense to clean everything up first.
What triggers the module to be updated when we update py-allspice code?

kdumontnu commented:
To publish we tag and push with an updated semantic version.

danielallspice committed:
Depricate for py-allspice

This was a wrapper / convenience function, but it complicates the dependencies too much.

danielallspice committed:
Depricate for py-allspice

Not needed anymore

danielallspice committed:
Depricate and rewrite for py-allspice

Rewrite on a different branch.

danielallspice committed:
Remove as incomplete

JSON bug + developer feedback.
Rewrite on a different branch.

danielallspice committed:
Add images for documentation

danielallspice committed:
Depricate for py-allspice

danielallspice committed:
Depricate for py-allspice

Rewrite on new branch.

danielallspice committed:
Depricate for py-allspice

Rewrite in new branch

danielallspice committed:
Depricate for py-allspice

Rewrite in new branch

danielallspice committed:
Add link to folder to readme.md

danielallspice committed:
Refactor for py-allspice

danielallspice committed:
Depricate for py-allspice

Rewrite tests in new branch

danielallspice committed:
Add install and use documentation to readme.md

danielallspice committed:
Add api example for getting sch/pcb json

danielallspice commented:
Added new example for allspice_generated/json/

danielallspice commented on GetRepoJSON/getRepoJSON.py:
```diff
@@ -0,0 +1,37 @@
+import os
+from gitea import *
+
+try:
```
@jtran , Can you take a look at this try/except block? This is in response to the log and raise. I still don't know if this is the way to do exception handling in python.

jtran replied:
Ideally, you only catch the type of exception you expect that you might hit.  So in this case:

```python
except KeyError:
```

jtran replied:
A linter like `ruff` would complain about this.

danielallspice replied:
Thank you, I wasn't sure which exception to use. I added KeyError

danielallspice committed:
Add documentation to remove busy wait

danielallspice committed:
Format with autopep8

danielallspice committed:
Create basic markdown file add

danielallspice committed:
Remove addfiletorepo from project

Not part of this branch

danielallspice committed:
Added note to remove rate limiting when ready

py-allspice rate limiting:
https://github.com/AllSpiceIO/py-allspice/issues/6

danielallspice commented:
@jtran Can you take a look at this PR and help me understand what I need to finish to get it through? Thank you!

jtran reviewed (approved):
👍

kdumontnu reviewed (commented):
Do we use those pngs anywhere?

danielallspice committed:
Remove unused images

danielallspice reviewed (commented):
@kdumontnu , I removed the unused image files. Good find!

jtran commented on GetRepoJSON/getRepoJSON.py:
```diff
@@ -0,0 +1,37 @@
+    URL   = os.environ['ALLSPICE_URL']
+except:
+    print("Invalid URL. Run this command:")
+    print("export ALLSPICE_URL=\"https://hub.allspice.io\"")
```
I think it's good that you're not printing and then re-raising, but this also continues running.  Is that intentional?  Do you maybe want to exit here?

danielallspice replied:
added quit() to both exception blocks.

jtran commented on GetRepoJSON/getRepoJSON.py:
```diff
@@ -0,0 +1,37 @@
+from time import sleep
+def delayserver():
+    sleep(0.01)
+    return
```
```suggestion
    return
```

This line is unneeded.

danielallspice committed:
Add ref param, exception type, autoformat

Refactoring:
changed delayserver to delay_server
added quit() to exception blocks
Fixed typo in exception message
